### PR TITLE
Replaces httpmock with mockito

### LINKED ISSUE #165: Hitting the fd limit again when running tests (aka Too many open files)
Depending on the OS, the file descriptor limit may be hit when running `teos` tests.

This is, in part, due to the tests in the `watchtower-plugin` that use `httpmock::MockServer` which spawns up to 25 servers when running the test suite. The issue can be patched in many ways, such as:

- Increasing the fd limit using `ulimit -n`
- Running the tests using a smaller `HTTPMOCK_MAX_SERVERS` (e.g.: `HTTPMOCK_MAX_SERVERS=10`)
- ...

In my personal case, I've gone for the first option, given that `macOS` has pretty default maxfiles (i.e. 256). This will become an issue if the fd limit is hit in other systems, like what happened with #119. We may like to consider an alternative option or even replace `httpmock::MockServer` with another mocking solution.

### PR DESCRIPTION
**Pull request by sr-gi**
This PR replaces our current mock for the API's HTTP interface ([httpmock](https://crates.io/crates/httpmock)) by [mockito](https://crates.io/crates/mockito)). The main reason for replacing one with the other is that the latter is more actively maintained and developed on, apart from it having some functionality that would be useful for us but the former lacked.

On top of that, `httpmock` was the root of some issues that we had worked around but were still carrying, such as #119 and #165.

It needs to be noted that, currently, we have replaced some of out time-based wait-and-check for the `Retrier` with some loops waiting for a condition to be met. This is due to `mockito` having some unexpected behavior for API delays when tests are run in parallel. Notice how this was also an issue with `httpmock`, even though we had the timings better nailed down (#193). We may be able to get rid of the loops once https://github.com/lipanski/mockito/issues/160 is fixed, even though it may be worth discussing whether wait-and-test is better than polling for a certain condition.

I've also spent some time redesigning the `Retrier` and the `RetryManager` to add some instrumentation points (such as adding a channel between the `WTClient` and the `Retrier` so the former can be notified when the state of the latter is changed). However, after several different iterations, I think it may not be worth the additional complexity for something that is merely used in testing (and can be checked with an ugly-but-simple loop)

### CODE REVIEW

**Comment by sr-gi** [comment]
I've added an additional commit improving the design given the devs from mockito added a new way of building mocks considering the request body when building the response (based on this issue I opened when I started working on the revamp: https://github.com/lipanski/mockito/issues/159).

Both commits can be squashed before merging, leaving them separate for now so it is easier to review. The delay issue seems to remain for now. If it happens to get fixed before we merge this I may add a last commit addressing it.

**Review by mariocynicys — COMMENTED** [review]
Only the 4 failing tests in #193 are patched with a `loop`, but I suspect other tests might fail too at some time.

note: The middle commit says `Fixes Carrier` but is actually for wt client retrier.

File: watchtower-plugin/src/retrier.rs
Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs:656]: I think it has to do with the mock having a synchronous wait. However, the function does not allow async closures atm.

I PRed a change to add delays to the repo but the maintainers seemed not to like it on the way it was (standalone, instead of trough something like this method), so I may rework it so async closures are accepted (I'm still debating about it).

**Comment by sr-gi** [comment]
This should do. I've ended up going for the loop alternative after seeing that even if the delays are properly working (as for the `.with_delay` method rejected by mockito) test can sometimes fail. wait-based tests with hardcoded times suck, especially when the underlying function is an exponential backoff. I'd be nice to have `with_body_from_fn` accept async closures so `tokio::time::sleep` can be used instead, but I'm honestly not fluent enough with async to do the change (nor do I think it's worth putting the time atm given the usecase).

I've added a macro to do the polling-based wait so things are less verbose. 

All commits can be squashed.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:604]: No need to export the macro.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:669]: We imported `std::thread` in the tests, let's omit the `std::` namespace.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:881]: Same as the one above.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:993]: Same as the one above the one above.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:718]: I guess this doesn't have to be in a nested scope? no?

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:582]: let's push this `use` down beside other `use teos_common`s.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:1188]: Also a redundant? scope here.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:710]: This should be analogous to `state.get_retrier_status(&tower_id).is_none()`, which we waited on above already.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:931]: Same as the one above.

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:1035]: Can we `wait_until!` the retrier is removed first and assert that it's misbehaving and not the other way around?
```rust
// Wait until the tower is no longer being retried.
wait_until!(wt_client
            .lock()
            .unwrap()
            .get_retrier_status(&tower_id)
            .is_none());

// Retriers are wiped every polling interval, so we'll need to wait a bit more to check it
tokio::time::sleep(Duration::from_secs_f64(1.5 * POLLING_TIME as f64)).await;

// The tower should have a misbehaving status.
assert!(wt_client
            .lock()
            .unwrap()
            .get_tower_status(&tower_id)
            .unwrap()
            .is_misbehaving());
```

File: watchtower-plugin/src/retrier.rs
Comment (mariocynicys) [inline watchtower-plugin/src/retrier.rs:1184]: Same here, we asserting on tower_id not in retriers but we already waited on it.

**Review by mariocynicys — CHANGES_REQUESTED** [review]
Some nits, looks good otherwise.

File: watchtower-plugin/src/retrier.rs
Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs:604]: True, this was originally intended to go on `teos-common`, but I changed my mind 😅

File: watchtower-plugin/src/retrier.rs
Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs:669]: That must have been vscode on its own, I'll remove the `std::thread` import since that'll create a smaller diff.

File: watchtower-plugin/src/retrier.rs
Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs:710]: Indeed

File: watchtower-plugin/src/retrier.rs
Comment (sr-gi) [reply watchtower-plugin/src/retrier.rs:1035]: We may even be able to get rid of the intermediate sleep in this case given the retrier should have already been wiped, wouldn't we?

Comment (mariocynicys) [reply watchtower-plugin/src/retrier.rs:1035]: > We may even be able to get rid of the intermediate sleep in this case given the retrier should have already been wiped, wouldn't we?

That's correct. The sleep was guarding against early assertion of `the retrier is not found`, so we can drop it.

**Comment by sr-gi** [comment]
Addressed all comments but the ones related to https://github.com/talaia-labs/rust-teos/pull/194#discussion_r1129241073

**Review by mariocynicys — APPROVED** [review]
lgtm

**Pull request merged**
